### DISCUSSION TITLE
Cannot use python script in custom target

### DISCUSSION THREAD

**Discussion by vpoinot** [mesonbuild/meson#13907] [Q&A] [answered]
I am actually trying to use the `custom_target` mechanism to generate a source file, and I cannot get it work...
Some details:

- I have a subproject (sqlpp11) that has a meson.build with the line:
```
meson.override_find_program('sqlite2cpp', files('scripts/sqlite2cpp.py'))
```
- in my main project's meson.build, I then have:
```
sqlite2cpp = find_program('sqlite2cpp', required: true)

schema = custom_target('generate-schema',
                       input: 'schema.sql',
                       output: 'schema.h',
                       command: [ sqlite2cpp, '-identity-naming', '@INPUT@', '@OUTPUT@', 'stox' ])

stox = executable('stox',
                  sources: [ 'main.cpp', schema ],
                  dependencies: boost)
```
- when I run meson, it does try to generate my schema.h file using my custom target, but it does this way:
```
[1/2] /home/vincent/devl/stox/subprojects/sqlpp11-0.64/scripts/sqlite2cpp.py -identity-naming ../../../schema.sql schema.h stox
FAILED: schema.h 
/home/vincent/devl/stox/subprojects/sqlpp11-0.64/scripts/sqlite2cpp.py -identity-naming ../../../schema.sql schema.h stox
/bin/sh: ligne 1: /home/vincent/devl/stox/subprojects/sqlpp11-0.64/scripts/sqlite2cpp.py: Permission denied
```
i.e. it does not try to use the interpreter python3 which is present in the shebang of the  sqlite2cpp.py script in this form (which, from what I understand, is understood by meson):
```
#!/usr/bin/env python3
```
Note that the sqlite2cpp.py script has its executable bit unset (as the "permission denied" error proves it), which should trigger the execution of the script through the interpreter.

(I am on Linux/Fedora, using meson 1.4.1, also tried meson 1.6.0, same)

Am I doing something wrong? or is there a bug somewhere?

**Comment by andy5995** [comment]
Maybe you need something like:

```
pymod = import('python')

py3 = pymod.find_installation('python3', required: true)

script = configure_file(...)
```

mentioned at the top of https://github.com/mesonbuild/meson/discussions/10978 ?

**Comment by eli-schwartz** [comment, marked as answer]
> i.e. it does not try to use the interpreter python3 which is present in the shebang of the sqlite2cpp.py script in this form (which, from what I understand, is understood by meson):
> 
> ```
> #!/usr/bin/env python3
> ```
> 
> Note that the sqlite2cpp.py script has its executable bit unset (as the "permission denied" error proves it), which should trigger the execution of the script through the interpreter.
> 
> (I am on Linux/Fedora, using meson 1.4.1, also tried meson 1.6.0, same)
> 
> Am I doing something wrong? or is there a bug somewhere?

Meson understands interpreter shebangs for scripts without the executable bit set, when going through find_program. More specifically, it's at the point where find_program *finds* a program that it parses this information. In your case, you use `files()` to pass to override_find_program, so future calls to `find_program()` detect this cached lookup.

`override_find_program()` when used with a `files()` object thinks that it got all the information it needs and doesn't bother checking for a shebang. Maybe it should, maybe it shouldn't, but the point is it certainly doesn't do so today.

But what does work is replacing:

```meson
meson.override_find_program('sqlite2cpp', files('scripts/sqlite2cpp.py'))
```

with:
```meson
meson.override_find_program('sqlite2cpp', find_program('scripts/sqlite2cpp.py'))
```
